components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/DigestAuthMediator.java

```diff
@@ -0,0 +1,501 @@
+                }
+
+                //get the Http method (GET, POST, PUT or DELETE etc.)
+                String httpMethod = (String) messageContext.getProperty(DigestAuthConstants.HTTP_METHOD);
+
+                if (log.isDebugEnabled()) {
+                    log.debug("HTTP method of request is : " + httpMethod);
+                }
+
+                //generate clientNonce
+                String clientNonce = generateClientNonce();
+
+                //calculate hash1
+                String ha1 = calculateHA1(userName, realm, passWord, algorithm, serverNonce, clientNonce);
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Value of hash 1 is : " + ha1);
+                }
+
+                //calculate hash2
+                String ha2 = calculateHA2(qop, httpMethod, digestUri, messageContext);
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Value of hash 2 is : " + ha2);
+                }
+
+                //getting the previous NonceCount
+                String prevNonceCount = (String) messageContext.getProperty(DigestAuthConstants.NONCE_COUNT);
+
+                if (prevNonceCount == null) {
+                    messageContext.setProperty(DigestAuthConstants.NONCE_COUNT, DigestAuthConstants.INIT_NONCE_COUNT);
+                    prevNonceCount = (String) messageContext.getProperty(DigestAuthConstants.NONCE_COUNT);
+                }
+
+                //generate the final hash (serverResponse)
+                String[] serverResponseArray = generateResponseHash(ha1, ha2, serverNonce, qop, prevNonceCount,
+                        clientNonce);
+
+                //setting the NonceCount after incrementing
+                messageContext.setProperty(DigestAuthConstants.NONCE_COUNT, serverResponseArray[1]);
+
+                String serverResponse = serverResponseArray[0];
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Value of server response  is : " + serverResponse);
+                }
+
+                //Construct the authorization header
+                StringBuilder header = constructAuthHeader(userName, realm, serverNonce, digestUri, serverResponseArray,
+                        qop, opaque, clientNonce, algorithm);
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Processed Authorization header to be sent in the request is : " + header);
+                }
+
+                //set the AuthHeader field in the message context
+                messageContext.setProperty(DigestAuthConstants.AUTH_HEADER, header.toString());
+
+                return true;
+
+            } else {
+                //This is not digest auth protected api. let it go. Might be basic auth or NTLM protected.
+                //Here we receive a www-authenticate header but it is not for Digest authentication.
+                return true;
+            }
+        } catch (NullPointerException ex) {
```

Comment:
Its might not be a very good idea to catch a NullPointerException as it can occur due to different reasons; not only "The endpoint does not support digest authentication". Can you please check whether there are any other alternative to handle this.